tsdb/head.go

```diff
@@ -1419,11 +1443,14 @@ func (a *headAppender) Append(ref uint64, lset labels.Labels, t int64, v float64
 // AppendExemplar for headAppender assumes the series ref already exists, and so it doesn't
 // use getOrCreate or make any of the lset sanity checks that Append does.
 func (a *headAppender) AppendExemplar(ref uint64, _ labels.Labels, e exemplar.Exemplar) (uint64, error) {
+	switch a.head.exemplars.(type) {
+	case noopExemplarStorage:
+		return 0, nil
+	}
 	// Check if exemplar storage is enabled.
-	if a.head.opts.NumExemplars <= 0 {
+	if !a.head.opts.EnableExemplarStorage && a.head.opts.MaxExemplars > -1 {
```

Comment:
I guess you meant this? (but look at my other comment about using `opts.MaxExemplars`)
```suggestion
	if !a.head.opts.EnableExemplarStorage || a.head.opts.MaxExemplars < 0 {
```